File: sudoku_generator.java
// sudoku_generator.java — Java версия

import java.io.*;
import java.util.*;

public class sudoku_generator {
    private int size = 9;
    private int boxSize = 3;
    private int[][] board;
    private int[][] solution;
    private int cellsToRemove;
    private Random rand;

    public sudoku_generator(String difficulty) {
        this.board = new int[size][size];
        this.solution = new int[size][size];
        this.rand = new Random();
        this.cellsToRemove = difficulty.equals("easy") ? 35 : difficulty.equals("medium") ? 45 : 55;
    }

    public void generate() {
        solve(board);
        for (int i = 0; i < size; i++) {
            System.arraycopy(board[i], 0, solution[i], 0, size);
        }
        removeCells();
    }

    private boolean solve(int[][] board) {
        int[] empty = findEmpty(board);
        if (empty == null) return true;
        int row = empty[0], col = empty[1];
        List<Integer> nums = new ArrayList<>();
        for (int i = 1; i <= 9; i++) nums.add(i);
        Collections.shuffle(nums, rand);
        for (int num : nums) {
            if (isValid(board, row, col, num)) {
                board[row][col] = num;
                if (solve(board)) return true;
                board[row][col] = 0;
            }
        }
        return false;
    }

    private int[] findEmpty(int[][] board) {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (board[i][j] == 0) return new int[]{i, j};
            }
        }
        return null;
    }

    private boolean isValid(int[][] board, int row, int col, int num) {
        for (int j = 0; j < size; j++) {
            if (board[row][j] == num) return false;
        }
        for (int i = 0; i < size; i++) {
            if (board[i][col] == num) return false;
        }
        int startRow = (row / boxSize) * boxSize;
        int startCol = (col / boxSize) * boxSize;
        for (int i = startRow; i < startRow + boxSize; i++) {
            for (int j = startCol; j < startCol + boxSize; j++) {
                if (board[i][j] == num) return false;
            }
        }
        return true;
    }

    private void removeCells() {
        List<int[]> cells = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                cells.add(new int[]{i, j});
            }
        }
        Collections.shuffle(cells, rand);
        int removed = 0;
        for (int[] cell : cells) {
            if (removed >= cellsToRemove) break;
            board[cell[0]][cell[1]] = 0;
            removed++;
        }
    }

    public void printBoard(boolean showSolution) {
        int[][] boardToPrint = showSolution ? solution : board;
        System.out.println("┌───────┬───────┬───────┐");
        for (int i = 0; i < size; i++) {
            System.out.print("│");
            for (int j = 0; j < size; j++) {
                if (boardToPrint[i][j] == 0) {
                    System.out.print(" . ");
                } else {
                    System.out.print(" " + boardToPrint[i][j] + " ");
                }
                if (j % 3 == 2 && j < size - 1) System.out.print("│");
            }
            System.out.println("│");
            if (i % 3 == 2 && i < size - 1) {
                System.out.println("├───────┼───────┼───────┤");
            }
        }
        System.out.println("└───────┴───────┴───────┘");
    }

    public void saveTXT(String filename) throws IOException {
        try (FileWriter fw = new FileWriter(filename)) {
            for (int[] row : board) {
                for (int j = 0; j < row.length; j++) {
                    fw.write((row[j] == 0 ? "." : String.valueOf(row[j])) + (j < row.length - 1 ? " " : ""));
                }
                fw.write("\n");
            }
        }
        System.out.println("💾 Сохранено: " + filename);
    }

    public void saveJSON(String filename) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"puzzle\":[");
        for (int i = 0; i < board.length; i++) {
            sb.append("[");
            for (int j = 0; j < board[i].length; j++) {
                sb.append(board[i][j]);
                if (j < board[i].length - 1) sb.append(",");
            }
            sb.append("]");
            if (i < board.length - 1) sb.append(",");
        }
        sb.append("],\"solution\":[");
        for (int i = 0; i < solution.length; i++) {
            sb.append("[");
            for (int j = 0; j < solution[i].length; j++) {
                sb.append(solution[i][j]);
                if (j < solution[i].length - 1) sb.append(",");
            }
            sb.append("]");
            if (i < solution.length - 1) sb.append(",");
        }
        sb.append("],\"difficulty\":\"easy\"}");
        try (FileWriter fw = new FileWriter(filename)) {
            fw.write(sb.toString());
        }
        System.out.println("💾 Сохранено: " + filename);
    }

    public static void main(String[] args) throws IOException {
        System.out.println("🧩 Sudoku Generator (Easy) (Java)");
        sudoku_generator gen = new sudoku_generator("easy");
        gen.generate();

        System.out.println("\nГоловоломка:");
        gen.printBoard(false);

        Scanner scanner = new Scanner(System.in);
        System.out.print("\nПоказать решение? (y/n): ");
        if (scanner.nextLine().trim().equalsIgnoreCase("y")) {
            System.out.println("\nРешение:");
            gen.printBoard(true);
        }

        gen.saveTXT("sudoku_puzzle.txt");
        gen.saveJSON("sudoku_puzzle.json");
        scanner.close();
    }
}

File: sudoku_generator.php
<?php
// sudoku_generator.php — PHP версия

class SudokuGenerator {
    private $size = 9;
    private $boxSize = 3;
    private $board = [];
    private $solution = [];
    private $cellsToRemove;

    public function __construct($difficulty = 'easy') {
        $this->board = array_fill(0, $this->size, array_fill(0, $this->size, 0));
        $this->solution = array_fill(0, $this->size, array_fill(0, $this->size, 0));
        $this->cellsToRemove = $difficulty == 'easy' ? 35 : ($difficulty == 'medium' ? 45 : 55);
    }

    public function generate() {
        $this->solve($this->board);
        for ($i = 0; $i < $this->size; $i++) {
            for ($j = 0; $j < $this->size; $j++) {
                $this->solution[$i][$j] = $this->board[$i][$j];
            }
        }
        $this->removeCells();
        return $this->board;
    }

    private function solve(&$board) {
        $empty = $this->findEmpty($board);
        if ($empty === null) return true;
        list($row, $col) = $empty;
        $nums = range(1, 9);
        shuffle($nums);
        foreach ($nums as $num) {
            if ($this->isValid($board, $row, $col, $num)) {
                $board[$row][$col] = $num;
                if ($this->solve($board)) return true;
                $board[$row][$col] = 0;
            }
        }
        return false;
    }

    private function findEmpty($board) {
        for ($i = 0; $i < $this->size; $i++) {
            for ($j = 0; $j < $this->size; $j++) {
                if ($board[$i][$j] == 0) return [$i, $j];
            }
        }
        return null;
    }

    private function isValid($board, $row, $col, $num) {
        for ($j = 0; $j < $this->size; $j++) {
            if ($board[$row][$j] == $num) return false;
        }
        for ($i = 0; $i < $this->size; $i++) {
            if ($board[$i][$col] == $num) return false;
        }
        $startRow = intdiv($row, $this->boxSize) * $this->boxSize;
        $startCol = intdiv($col, $this->boxSize) * $this->boxSize;
        for ($i = $startRow; $i < $startRow + $this->boxSize; $i++) {
            for ($j = $startCol; $j < $startCol + $this->boxSize; $j++) {
                if ($board[$i][$j] == $num) return false;
            }
        }
        return true;
    }

    private function removeCells() {
        $cells = [];
        for ($i = 0; $i < $this->size; $i++) {
            for ($j = 0; $j < $this->size; $j++) {
                $cells[] = [$i, $j];
            }
        }
        shuffle($cells);
        $removed = 0;
        foreach ($cells as $cell) {
            if ($removed >= $this->cellsToRemove) break;
            $this->board[$cell[0]][$cell[1]] = 0;
            $removed++;
        }
    }

    public function printBoard($showSolution = false) {
        $board = $showSolution ? $this->solution : $this->board;
        echo "┌───────┬───────┬───────┐\n";
        for ($i = 0; $i < $this->size; $i++) {
            echo "│";
            for ($j = 0; $j < $this->size; $j++) {
                if ($board[$i][$j] == 0) {
                    echo " . ";
                } else {
                    echo " {$board[$i][$j]} ";
                }
                if ($j % 3 == 2 && $j < $this->size - 1) echo "│";
            }
            echo "│\n";
            if ($i % 3 == 2 && $i < $this->size - 1) {
                echo "├───────┼───────┼───────┤\n";
            }
        }
        echo "└───────┴───────┴───────┘\n";
    }

    public function saveTXT($filename = 'sudoku_puzzle.txt') {
        $content = "";
        foreach ($this->board as $row) {
            $content .= implode(' ', array_map(function($c) { return $c == 0 ? '.' : $c; }, $row)) . "\n";
        }
        file_put_contents($filename, $content);
        echo "💾 Сохранено: $filename\n";
    }

    public function saveJSON($filename = 'sudoku_puzzle.json') {
        $data = ['puzzle' => $this->board, 'solution' => $this->solution, 'difficulty' => 'easy'];
        file_put_contents($filename, json_encode($data, JSON_PRETTY_PRINT));
        echo "💾 Сохранено: $filename\n";
    }
}

function main() {
    echo "🧩 Sudoku Generator (Easy) (PHP)\n";
    $gen = new SudokuGenerator('easy');
    $gen->generate();

    echo "\nГоловоломка:\n";
    $gen->printBoard();

    echo "\nПоказать решение? (y/n): ";
    $ans = trim(fgets(STDIN));
    if (strtolower($ans) == 'y') {
        echo "\nРешение:\n";
        $gen->printBoard(true);
    }

    $gen->saveTXT();
    $gen->saveJSON();
}

main();
?>
